### PR TITLE
Implement use case architecture

### PR DESCRIPTION
**Pull request by sven1103**

### CODE REVIEW

File: domain/src/main/java/life/qbic/usermanagement/User.java [hunk 1]
```diff
@@ -6,11 +6,12 @@
...
 import javax.persistence.Table;
+import jdk.jshell.spi.ExecutionControl.UserException;
 import life.qbic.usermanagement.policies.*;
 
 /**
  * <b>User class</b>
- *
+ * <p>
  * User aggregate in the context of user management.
  *
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/User.java:16]:
```suggestion
 * </p>
```

File: domain/src/main/java/life/qbic/usermanagement/registration/RegisterUserInput.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+ * Input interface to register a new user in the application.
+ *
+ * @since 1.0.0
+ */
+public interface RegisterUserInput {
+
+  void register(User user);
+
+  void setOutput(RegisterUserOutput output);
+
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/registration/RegisterUserInput.java:15]: javadocs

File: domain/src/main/java/life/qbic/usermanagement/registration/RegisterUserOutput.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package life.qbic.usermanagement.registration;
+
+public interface RegisterUserOutput {
+
+  void onSuccess();
+
+  void onFailure(String reason);
+
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/registration/RegisterUserOutput.java:8]: missing javadoc for method and class

Comment (jenniferboedker) [reply domain/src/main/java/life/qbic/usermanagement/registration/RegisterUserOutput.java:8]: *interface

File: domain/src/main/java/life/qbic/usermanagement/registration/Registration.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+  @Override
+  public void register(User user) {
+    userRepository.findByEmail(user.getEmail())
+        .ifPresentOrElse(
+            u -> registerUserOutput.onFailure("User with email address already exists."),
+            () -> {
+              user.setEmailConfirmed(false);
+              userRepository.addUser(user);
+              registerUserOutput.onSuccess();
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/registration/Registration.java:39]: you might want to make sure, this was set properyl? Now the output could be null

File: domain/src/main/java/life/qbic/usermanagement/registration/Registration.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  private RegisterUserOutput registerUserOutput;
+
+  private final UserRepository userRepository;
+
+  public Registration(UserRepository userRepository) {
+    this.userRepository = userRepository;
+  }
+
+  public void setRegisterUserOutput(RegisterUserOutput registerUserOutput) {
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/registration/Registration.java:27]: java doc missing

File: webapp/frontend/generated/theme.d.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+export declare const applyTheme: (target: Node) => void;
```
Comment (jenniferboedker) [inline webapp/frontend/generated/theme.d.ts:1]:
```suggestion
export declare const applyTheme: (target: Node) => void;

```

File: webapp/src/main/java/life/qbic/AppConfig.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
```
Comment (jenniferboedker) [inline webapp/src/main/java/life/qbic/AppConfig.java:18]: missing doc

File: webapp/src/main/java/life/qbic/views/register/RegisterHandler.java [hunk 8]
```diff
@@ -3,71 +3,90 @@
...
+    this.registerUserInput = registerUserInput;
+    registerUserInput.setOutput(this);
   }
 
   @Override
   public boolean register(RegisterLayout registerLayout) {
-    if (registeredRegisterLayout != registerLayout) {
-      this.registeredRegisterLayout = registerLayout;
+    if (userRegistrationLayout != registerLayout) {
+      this.userRegistrationLayout = registerLayout;
```
Comment (jenniferboedker) [inline webapp/src/main/java/life/qbic/views/register/RegisterHandler.java:36]: I think by this renaming its not clear what the difference between the two layouts is. One was already registered in the layout, the other not

File: webapp/src/main/java/life/qbic/views/register/RegisterHandler.java [hunk 9]
```diff
@@ -3,71 +3,90 @@
...
+  @Override
+  public void onSuccess() {
+    UI.getCurrent().navigate("/login");
+  }
+
+  @Override
+  public void onFailure(String reason) {
+    // Display error to the user
+    // Stub output:
+    System.out.println(reason);
```
Comment (jenniferboedker) [inline webapp/src/main/java/life/qbic/views/register/RegisterHandler.java:89]: you might want to use logging here?

File: domain/src/main/java/life/qbic/usermanagement/User.java (hunk 1 above)
Comment (sven1103) [reply domain/src/main/java/life/qbic/usermanagement/User.java:16]: That's how its done when using the auto refactor from IntelliJ. We should make use of it and not fight against the tools

Comment (jenniferboedker) [reply domain/src/main/java/life/qbic/usermanagement/User.java:16]: but is it then properly rendered?

Comment (jenniferboedker) [reply domain/src/main/java/life/qbic/usermanagement/User.java:16]: I think we need the closing tag here

File: domain/src/main/java/life/qbic/usermanagement/registration/Registration.java (hunk 4 above)
Comment (sven1103) [reply domain/src/main/java/life/qbic/usermanagement/registration/Registration.java:39]: yes, tbh I dont like how it is currently set up to break the cyclic dependency. A use case without an output is just wrong by design. 

I dont have a better idea though how to solve it.

File: domain/src/main/java/life/qbic/usermanagement/User.java (hunk 1 above)
Comment (sven1103) [reply domain/src/main/java/life/qbic/usermanagement/User.java:16]: not needed, you can check out and see for yourself :)

File: webapp/frontend/generated/theme.d.ts (hunk 6 above)
Comment (sven1103) [reply webapp/frontend/generated/theme.d.ts:1]: not my code 🤷🏽

File: domain/src/main/java/life/qbic/usermanagement/registration/Registration.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+      @Override
+      public void onFailure(String reason) {
+        System.err.println("Called dummy register failure output.");
+      }
+    };
+  }
+
+  /**
+   * Sets an overrides the use case output.
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/usermanagement/registration/Registration.java:51]:
```suggestion
   * Sets and overrides the use case output.
```

File: webapp/src/main/java/life/qbic/views/register/RegisterHandler.java (hunk 8 above)
Comment (sven1103) [reply webapp/src/main/java/life/qbic/views/register/RegisterHandler.java:36]: I am confused with too many `registered` here. I think I will do another round of renaming to make it more explicit.

File: webapp/src/main/java/life/qbic/views/register/RegisterHandlerInterface.java [hunk 11]
```diff
@@ -1,17 +1,17 @@
 package life.qbic.views.register;
 
 /**
- * <b> Interface to register the {@link RegisterLayout} to the {@link RegisterHandler}. </b>
+ * <b> Interface to register the {@link UserRegistrationLayout} to the {@link UserRegistrationHandler}. </b>
  *
  * @since 1.0.0
  */
 public interface RegisterHandlerInterface {
```
Comment (jenniferboedker) [inline webapp/src/main/java/life/qbic/views/register/RegisterHandlerInterface.java:8]: rename of this class to make it consistent

Comment (jenniferboedker) [reply webapp/src/main/java/life/qbic/views/register/RegisterHandlerInterface.java:8]:
```suggestion
public interface UserRegistrationHandlerInterface {
```

File: webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+  @Autowired
+  UserRegistrationHandler(RegisterUserInput registrationUseCase) {
+    this.registrationUseCase = registrationUseCase;
+    registrationUseCase.setOutput(this);
+  }
+
+  @Override
+  public boolean register(UserRegistrationLayout registrationLayout) {
+    if (userRegistrationLayout != registrationLayout) {
```
Comment (jenniferboedker) [inline webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java:35]:
```suggestion
    if (registeredUserRegistrationLayout != registrationLayout) {
```

File: webapp/src/main/java/life/qbic/views/MainHandler.java [hunk 13]
```diff
@@ -32,7 +32,7 @@ private void addClickListeners() {
 
     registeredMainLayout.register.addClickListener(
         event -> {
-          UI.getCurrent().navigate("register");
+          UI.getCurrent().navigate("handle");
```
Comment (jenniferboedker) [inline webapp/src/main/java/life/qbic/views/MainHandler.java:35]: thats not right, the route does not change!

Comment (jenniferboedker) [reply webapp/src/main/java/life/qbic/views/MainHandler.java:35]:
```suggestion
          UI.getCurrent().navigate("register");
```

File: webapp/src/main/java/life/qbic/views/register/UserRegistrationLayout.java [hunk 14]
```diff
@@ -24,12 +24,12 @@
...
 
 /**
- * <b> Defines the look of the register layout. </b>
+ * <b> Defines the look of the handle layout. </b>
  *
  * @since 1.0.0
  */
 @PageTitle("Register")
-@Route(value = "register", layout = MainLayout.class)
+@Route(value = "handle", layout = MainLayout.class)
```
Comment (jenniferboedker) [inline webapp/src/main/java/life/qbic/views/register/UserRegistrationLayout.java:32]: why?
Its still register for the use case

File: webapp/src/main/java/life/qbic/views/register/UserRegistrationLayout.java [hunk 15]
```diff
@@ -52,7 +52,7 @@ public class UserRegistrationLayout extends VerticalLayout {
   private H3 layoutTitle;
 
   public UserRegistrationLayout(@Autowired UserRegistrationHandlerInterface registerHandler) {
-    setId("register-view");
+    setId("handle-view");
```
Comment (jenniferboedker) [inline webapp/src/main/java/life/qbic/views/register/UserRegistrationLayout.java:55]: change it back pls, this breaks the styling!

Comment (sven1103) [reply webapp/src/main/java/life/qbic/views/register/UserRegistrationLayout.java]: sorry, done

File: webapp/src/main/java/life/qbic/views/register/UserRegistrationLayout.java [hunk 16]
```diff
@@ -24,15 +24,15 @@
 import java.util.stream.Stream;
 
 /**
- * <b> Defines the look of the register layout. </b>
+ * <b> Defines the look of the handle layout. </b>
```
Comment (jenniferboedker) [inline webapp/src/main/java/life/qbic/views/register/UserRegistrationLayout.java:27]:
```suggestion
 * <b> Defines the look of the registration layout. </b>
```

**Review by jenniferboedker — APPROVED** [review]

**Pull request merged**